### PR TITLE
Improve `admin_destroy_market` and `admin_move_market_to_resolved` benchmarks

### PR DESCRIPTION
**Pull request by Chralt98**

### CODE REVIEW

**Comment by Chralt98** [comment]
Blocked until https://github.com/zeitgeistpm/zeitgeist/pull/806 is merged.

**Comment by maltekliemann** [comment]
Starting my review now...

File: zrml/prediction-markets/src/benchmarks.rs [hunk 1]
```diff
@@ -230,29 +234,155 @@ fn setup_resolve_common_scalar_after_dispute<T: Config>(
...
 
         let destroy_origin = T::DestroyOrigin::successful_origin();
         let call = Call::<T>::admin_destroy_market { market_id };
-    }: { call.dispatch_bypass_filter(destroy_origin)? }
+    }: {
+        call.dispatch_bypass_filter(destroy_origin)?
+    } verify {
+        assert_last_event::<T>(Event::MarketDestroyed::<T>(
+            market_id,
+        ).into());
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/benchmarks.rs:334]:
```suggestion
        assert_last_event::<T>(Event::MarketDestroyed::<T>(market_id).into());
```

File: zrml/prediction-markets/src/lib.rs [hunk 2]
```diff
@@ -208,10 +256,28 @@ mod pallet {
...
             );
-            Self::clear_auto_resolve(&market_id)?;
+            let (ids_len, disputes_len) = Self::clear_auto_resolve(&market_id)?;
             let market = T::MarketCommons::market(&market_id)?;
-            let weight = Self::on_resolution(&market_id, &market)?;
-            Ok((Some(weight), Pays::No).into())
+            let weight_on_resolution = Self::on_resolution(&market_id, &market)?;
+            // Substracted by internal_resolve_scalar_...,
+            // because it's returned in `on_resolution` and the benchmark of
+            // `admin_move_market_to_resolved_overhead_...` was initialised with a scalar market
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/lib.rs:264]: I'm very concerned about this approach. This seems hard to maintain, hard to understand and makes the code difficult to read. There are two alternatives:

- Benchmark each function like `admin_move_market_to_resolved` by identifying each of its execution paths and benchmarking each path separately
- Use dependency injection to inject a noop `on_resolution`

Both are fine in my mind, although the latter seems to subordinate the structure of the code to the benchmarks, which is why I would actually prefer the first approach.

File: zrml/prediction-markets/src/benchmarks.rs [hunk 3]
```diff
@@ -230,29 +234,155 @@ fn setup_resolve_common_scalar_after_dispute<T: Config>(
...
+        MarketCreation::Permissionless,
+        MarketType::Categorical(categories.saturated_into()),
+        ScoringRule::CPMM,
+        Some(MarketPeriod::Timestamp(range_start..range_end)),
+    )?;
+
+    let max_swap_fee: BalanceOf<T> = MaxSwapFee::get().saturated_into();
+    let min_liquidity: BalanceOf<T> = MinLiquidity::get().saturated_into();
+    let _ = Call::<T>::buy_complete_set { market_id, amount: min_liquidity }
+        .dispatch_bypass_filter(RawOrigin::Signed(caller.clone()).into())?;
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/prediction-markets/src/benchmarks.rs]: 2e6721ebd07bc79885337cfb0e0a48b08e2732d0

File: zrml/prediction-markets/src/benchmarks.rs [hunk 4]
```diff
@@ -230,29 +234,155 @@ fn setup_resolve_common_scalar_after_dispute<T: Config>(
...
+    let weight_len: usize = MaxRuntimeUsize::from(categories).into();
+    let weights = vec![MinWeight::get(); weight_len];
+    Pallet::<T>::deploy_swap_pool_for_market(
+        RawOrigin::Signed(caller.clone()).into(),
+        market_id,
+        max_swap_fee,
+        min_liquidity,
+        weights,
+    )
+    .unwrap();
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/prediction-markets/src/benchmarks.rs]: 37d8381cf46d14464ef870b8b6e3266337278ad2

File: zrml/prediction-markets/src/benchmarks.rs [hunk 5]
```diff
@@ -230,29 +234,155 @@ fn setup_resolve_common_scalar_after_dispute<T: Config>(
...
+
 benchmarks! {
     admin_destroy_disputed_market{
-        let categories = T::MaxCategories::get();
-        let (caller, market_id) = setup_resolve_common_categorical::<T>(0, 0, categories)?;
+        let a in (T::MinCategories::get().into())..T::MaxCategories::get().into();
+        let d in 1..T::MaxDisputes::get();
+        let o in 0..63;
+        let c in 0..63;
+        let r in 0..63;
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/prediction-markets/src/benchmarks.rs]: Wow, nice suggestion, because we struggle to know the purpose of this components in the future!

 2c60636c8cee1f361935065554e836091bf03db5

File: zrml/prediction-markets/src/benchmarks.rs [hunk 6]
```diff
@@ -230,29 +234,155 @@ fn setup_resolve_common_scalar_after_dispute<T: Config>(
...
+        let (caller, market_id) = create_reported_market_with_pool::<T>(a)?;
+
+        let pool_id = T::MarketCommons::market_pool(&market_id).unwrap();
+
+        // minimum of number of disputes and (categories - 1) because of `matches_outcome_report`
+        for i in 1..=(a - 1).min(d) {
+            let outcome = OutcomeReport::Categorical(i.saturated_into());
+            let disputor = account("disputor", i, 0);
+            let dispute_bond = T::DisputeBond::get() +
+                (T::DisputeFactor::get() * i.saturated_into::<u32>().into());
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/prediction-markets/src/benchmarks.rs]: 5f00ae1c86a6cded4dcf892b91883d145a76f03e

File: zrml/prediction-markets/src/benchmarks.rs [hunk 7]
```diff
@@ -230,29 +234,155 @@ fn setup_resolve_common_scalar_after_dispute<T: Config>(
...
-            let disputes = crate::Disputes::<T>::get(market_id);
-            let market = T::MarketCommons::market(&Default::default()).unwrap();
-            T::SimpleDisputes::on_dispute(&disputes, &market_id, &market)?;
+        let (caller, market_id) = create_reported_market_with_pool::<T>(a)?;
+
+        let pool_id = T::MarketCommons::market_pool(&market_id).unwrap();
+
+        // minimum of number of disputes and (categories - 1) because of `matches_outcome_report`
+        for i in 1..=(a - 1).min(d) {
+            let outcome = OutcomeReport::Categorical(i.saturated_into());
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/prediction-markets/src/benchmarks.rs:292]: Did you take the number of disputes into account with this suggestion? I mean it could be, that there is one dispute, but `a = 5` and then I would get the `matches_outcome_report` produced error.

File: zrml/prediction-markets/src/benchmarks.rs [hunk 8]
```diff
@@ -230,29 +234,155 @@ fn setup_resolve_common_scalar_after_dispute<T: Config>(
...
+
+        let disputes = Disputes::<T>::get(market_id);
+        // TODO(#730): MarketIdsPerDisputeBlock will store the future block number.
+        let last_dispute = disputes.last().unwrap();
+        let dispute_at = last_dispute.at;
+        for i in 0..r {
+            MarketIdsPerDisputeBlock::<T>::try_mutate(
+                dispute_at,
+                |ids| ids.try_push(i.into()),
+            ).unwrap();
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/prediction-markets/src/benchmarks.rs:328]: I am simulating the presence of other markets, yes. Because MarketIdsPerDisputeBlock has multiple ids inside it. I think the benchmark component comments for that are sufficient? 

 2c60636c8cee1f361935065554e836091bf03db5

File: zrml/prediction-markets/src/lib.rs [hunk 9]
```diff
@@ -128,14 +138,15 @@ mod pallet {
                 &market_account,
                 T::AssetManager::free_balance(Asset::Ztg, &market_account),
             );
+            let mut category_count = 0u32;
             if let Ok(pool_id) = T::MarketCommons::market_pool(&market_id) {
-                T::Swaps::destroy_pool(pool_id)?;
+                (_, category_count) = T::Swaps::destroy_pool(pool_id)?;
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/prediction-markets/src/lib.rs]: You are right, it's totally strange to do it this way I was doing it. You are again right, that the storage overlay of Substrate does take care of it. So I didn't use the argument passing.

2834b11444bff599adcd776e9c873ca8d0456f05

File: zrml/prediction-markets/src/lib.rs [hunk 10]
```diff
@@ -1523,27 +1589,38 @@ mod pallet {
...
-                });
-            }
-            if market.status == MarketStatus::Disputed {
-                let disputes = Disputes::<T>::get(market_id);
-                if let Some(last_dispute) = disputes.last() {
-                    let at = last_dispute.at;
-                    let mut old_disputes_per_block = MarketIdsPerDisputeBlock::<T>::get(at);
-                    remove_item::<MarketIdOf<T>, _>(&mut old_disputes_per_block, market_id);
-                    MarketIdsPerDisputeBlock::<T>::mutate(at, |ids| {
+            let mut disputes_len = 0u32;
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/prediction-markets/src/lib.rs]: f37b69966ec53261e5a2e487ad74f481faa9accd

File: zrml/prediction-markets/src/lib.rs (hunk 2 above)
Comment (Chralt98) [reply zrml/prediction-markets/src/lib.rs:269]: Ok, but then I would like to wait for https://github.com/zeitgeistpm/zeitgeist/pull/813 to get merged.

File: zrml/prediction-markets/src/benchmarks.rs [hunk 11]
```diff
@@ -230,29 +234,155 @@ fn setup_resolve_common_scalar_after_dispute<T: Config>(
...
-        let (caller, market_id) = setup_resolve_common_categorical::<T>(0, 0, categories)?;
         let destroy_origin = T::DestroyOrigin::successful_origin();
         let call = Call::<T>::admin_destroy_market { market_id };
-    }: { call.dispatch_bypass_filter(destroy_origin)? }
+    }: {
+        call.dispatch_bypass_filter(destroy_origin)?
+    } verify {
+        assert_last_event::<T>(Event::MarketDestroyed::<T>(
+            market_id,
+        ).into());
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/prediction-markets/src/benchmarks.rs:390]: 3d4e7f7b0805bb754fcb4f289d556771303e4c11

**Comment by Chralt98** [comment]
Waiting for the merge of https://github.com/zeitgeistpm/zeitgeist/pull/813 ..

File: zrml/prediction-markets/src/lib.rs (hunk 2 above)
Comment (Chralt98) [reply zrml/prediction-markets/src/lib.rs]: https://github.com/zeitgeistpm/zeitgeist/pull/812/commits/32f9ca233831fca3654894158335e9dce31c19c1

File: zrml/prediction-markets/src/benchmarks.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (vivekvpandya) [reply zrml/prediction-markets/src/benchmarks.rs:243]: just a question, since d represents number of disputes I think shouldn't we use that for iteration? why number of categories should influence disputes reported to market?

File: zrml/prediction-markets/src/benchmarks.rs [hunk 12]
```diff
@@ -276,27 +220,127 @@ benchmarks! {
...
+        // minimum of number of disputes and (categories - 1) because of `matches_outcome_report`
+        for i in 1..=(a - 1).min(d) {
+            let outcome = OutcomeReport::Categorical(i.saturated_into());
+            let disputor = account("disputor", i, 0);
+            let dispute_bond = crate::pallet::default_dispute_bond::<T>(i as usize);
+            T::AssetManager::deposit(Asset::Ztg, &disputor, dispute_bond)?;
+            let _ = Pallet::<T>::dispute(RawOrigin::Signed(disputor).into(), market_id, outcome)?;
+        }
+
+        let market = T::MarketCommons::market(&market_id.saturated_into())?;
```
Comment (vivekvpandya) [inline zrml/prediction-markets/src/benchmarks.rs:250]:
```suggestion
        let market = T::MarketCommons::market(&market_id)?;
```

File: zrml/prediction-markets/src/benchmarks.rs [hunk 13]
```diff
@@ -276,27 +220,127 @@ benchmarks! {
...
+        let r in 0..63;
+
+        let (caller, market_id) = setup_reported_categorical_market_with_pool::<T>(
+            a,
+            OutcomeReport::Categorical(0u16),
+        )?;
+
+        let pool_id = T::MarketCommons::market_pool(&market_id)?;
+
+        let market = T::MarketCommons::market(&market_id.saturated_into())?;
```
Comment (vivekvpandya) [inline zrml/prediction-markets/src/benchmarks.rs:306]:
```suggestion
        let market = T::MarketCommons::market(&market_id)?;
```

File: zrml/prediction-markets/src/benchmarks.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (vivekvpandya) [reply zrml/prediction-markets/src/benchmarks.rs:342]: @maltekliemann shouldn't the event be emitted only when market is properly destroyed?

File: zrml/prediction-markets/src/benchmarks.rs [hunk 14]
```diff
@@ -329,15 +373,180 @@ benchmarks! {
...
+        for i in 1..=d {
+            let outcome = OutcomeReport::Categorical((i % 2).saturated_into::<u16>());
+            let disputor = account("disputor", i, 0);
+            let dispute_bond = crate::pallet::default_dispute_bond::<T>(i as usize);
+            T::AssetManager::deposit(
+                Asset::Ztg,
+                &disputor,
+                dispute_bond,
+            )?;
+            let _ = Pallet::<T>::dispute(RawOrigin::Signed(disputor).into(), market_id, outcome)?;
```
Comment (vivekvpandya) [inline zrml/prediction-markets/src/benchmarks.rs:526]:
```suggestion
Pallet::<T>::dispute(RawOrigin::Signed(disputor).into(), market_id, outcome)?;
```

File: zrml/prediction-markets/src/benchmarks.rs [hunk 15]
```diff
@@ -329,15 +373,180 @@ benchmarks! {
...
+        for i in 1..=d {
+            let outcome = OutcomeReport::Scalar(i.saturated_into());
+            let disputor = account("disputor", i, 0);
+            let dispute_bond = crate::pallet::default_dispute_bond::<T>(i as usize);
+            T::AssetManager::deposit(
+                Asset::Ztg,
+                &disputor,
+                dispute_bond,
+            )?;
+            let _ = Pallet::<T>::dispute(RawOrigin::Signed(disputor).into(), market_id, outcome)?;
```
Comment (vivekvpandya) [inline zrml/prediction-markets/src/benchmarks.rs:475]:
```suggestion
Pallet::<T>::dispute(RawOrigin::Signed(disputor).into(), market_id, outcome)?;
```

File: zrml/prediction-markets/src/benchmarks.rs [hunk 16]
```diff
@@ -656,10 +868,10 @@ benchmarks! {
         let markets = BoundedVec::try_from(vec![market_info; a as usize]).unwrap();
         <MarketsCollectingSubsidy<T>>::put(markets);
     }: {
-        Pallet::<T>::process_subsidy_collecting_markets(
+        let _ = Pallet::<T>::process_subsidy_collecting_markets(
```
Comment (vivekvpandya) [inline zrml/prediction-markets/src/benchmarks.rs:871]: why this change is needed?

Comment (Chralt98) [reply zrml/prediction-markets/src/benchmarks.rs:871]:
```suggestion
        Pallet::<T>::process_subsidy_collecting_markets(
```

**Comment by maltekliemann** [comment]
Starting my review now.

File: zrml/prediction-markets/src/benchmarks.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/prediction-markets/src/benchmarks.rs:342]: Sure, just seemed strange to me to do it like this, but... yeah, I guess we can do it this way.

File: zrml/prediction-markets/src/benchmarks.rs [hunk 17]
```diff
@@ -329,15 +373,180 @@ benchmarks! {
...
+        for i in 1..=d {
+            let outcome = OutcomeReport::Scalar(i.saturated_into());
+            let disputor = account("disputor", i, 0);
+            let dispute_bond = crate::pallet::default_dispute_bond::<T>(i as usize);
+            T::AssetManager::deposit(
+                Asset::Ztg,
+                &disputor,
+                dispute_bond,
+            )?;
+Pallet::<T>::dispute(RawOrigin::Signed(disputor).into(), market_id, outcome)?;
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/benchmarks.rs:475]:
```suggestion
            Pallet::<T>::dispute(RawOrigin::Signed(disputor).into(), market_id, outcome)?;
```

File: zrml/prediction-markets/src/benchmarks.rs [hunk 18]
```diff
@@ -329,15 +373,180 @@ benchmarks! {
...
+        for i in 1..=d {
+            let outcome = OutcomeReport::Categorical((i % 2).saturated_into::<u16>());
+            let disputor = account("disputor", i, 0);
+            let dispute_bond = crate::pallet::default_dispute_bond::<T>(i as usize);
+            T::AssetManager::deposit(
+                Asset::Ztg,
+                &disputor,
+                dispute_bond,
+            )?;
+Pallet::<T>::dispute(RawOrigin::Signed(disputor).into(), market_id, outcome)?;
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/benchmarks.rs:526]:
```suggestion
            Pallet::<T>::dispute(RawOrigin::Signed(disputor).into(), market_id, outcome)?;
```

File: zrml/prediction-markets/src/lib.rs [hunk 19]
```diff
@@ -208,9 +260,28 @@ mod pallet {
...
+                    );
+                }
+            } else if let MarketStatus::Reported = market.status {
+                weight = T::WeightInfo::admin_move_market_to_resolved_scalar_reported(ids_len);
+            } else {
+                weight = T::WeightInfo::admin_move_market_to_resolved_scalar_disputed(
+                    ids_len,
+                    disputes_len,
+                );
+            }
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/lib.rs:284]: [nitpick] This is prone to error. It's not exactly clear that the `else` branch is for disputed scalar markets. When we add other market types, this might be a problem.

I think matching `market.market_type` and `market.status` ~~(and raising an error is the `status` is not as expected!)~~ is safer.

**Review by maltekliemann — DISMISSED** [review]
:thumbsup:

File: zrml/prediction-markets/src/lib.rs (hunk 19 above)
Comment (Chralt98) [reply zrml/prediction-markets/src/lib.rs]: 2447d4986b4d42182fd4edb88413d2decc69f4bf

File: zrml/prediction-markets/src/benchmarks.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/prediction-markets/src/benchmarks.rs]: f922fcf9e8b779451c67333d24875f6a8a1fd632

**Review by vivekvpandya — APPROVED** [review]
Thanks for this PR!

File: zrml/prediction-markets/src/lib.rs [hunk 20]
```diff
@@ -1647,28 +1725,43 @@ mod pallet {
...
+                            dispute_duration_ends_at_block,
+                            |ids| -> (u32, u32) {
+                                let ids_len = ids.len() as u32;
+                                remove_item::<MarketIdOf<T>, _>(ids, market_id);
+                                (ids_len, disputes.len() as u32)
+                            },
+                        )
+                    } else {
+                        (0u32, disputes.len() as u32)
+                    }
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/lib.rs:1759]: Sorry, but this is another edition of "I missed it in the first review". Are these cases possible? In other words, if the `market_status` is `Disputed`, won't `disputes.last()` always be `Some`? In that case, I think we should just use `disputes.last().ok_or(/* error */)` and get rid of the else branch.

File: zrml/prediction-markets/src/lib.rs [hunk 21]
```diff
@@ -1647,28 +1725,43 @@ mod pallet {
...
+                }
+                _ => (0u32, 0u32),
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/lib.rs:1761]: Sorry, but this is another edition of "I missed it in the first review". I think this branch **should** never happen, right? In that case, I would rather throw an error.

```suggestion
                _ => return Error::<T>::InvalidMarketStatus,
```

**Review by maltekliemann — APPROVED** [review]
Approved, but not without leaving another two nitpicks about the error handling. Sorry, should have mentioned those sooner.

File: zrml/prediction-markets/src/lib.rs (hunk 21 above)
Comment (Chralt98) [reply zrml/prediction-markets/src/lib.rs:1761]: In `admin_destroy_market` we also call this function `clear_auto_resolve` and I don't like to throw an error there, because it should be callable in any state.

File: zrml/prediction-markets/src/lib.rs (hunk 20 above)
Comment (Chralt98) [reply zrml/prediction-markets/src/lib.rs:1759]: In that case you are right.

File: zrml/prediction-markets/src/lib.rs (hunk 21 above)
Comment (maltekliemann) [reply zrml/prediction-markets/src/lib.rs:1760]: Oh, nice. :thumbsup:

**Review by maltekliemann — APPROVED** [review]
:thumbsup:
